### PR TITLE
Gatsby preview extensions

### PR DESCRIPTION
**Pull request by floelhoeffel**

### CODE REVIEW

**Comment by floelhoeffel** [comment]
@jelz @suevalov its not 100% done but functioning. I would appreciate an early review if someone has time.

File: samples/gatsby-preview/extension.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  "srcdoc": "./build/index.html",
+  "fieldTypes": ["Symbol"],
+  "sidebar": true,
+  "parameters": {
+    "instance": [
+      {
+        "id": "contentTypeSlug",
+        "type": "Symbol",
+        "name": "URL slug for content type",
+        "default": ""
```
Comment (jelz) [inline samples/gatsby-preview/extension.json]: this could be skipped

File: samples/gatsby-preview/src/index.jsx [hunk 2]
```diff
@@ -0,0 +1,165 @@
+class App extends React.Component {
+  constructor(props) {
+    super(props)
+
+    let parameters = this.props.sdk.parameters
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: Use `const`

File: samples/gatsby-preview/src/index.jsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    let projectId = parameters.installation.projectId
+    let contentTypeSlug = parameters.instance.contentTypeSlug
+    let isAutoUpdate = parameters.instance.isAutoUpdate
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: `const { contentTypeSlug, isAutoUpdate } = parameters.instance`

File: samples/gatsby-preview/src/index.jsx [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    this.state = {
+      projectId: projectId,
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: no need to repeat, just do `projectId,`

File: samples/gatsby-preview/src/index.jsx [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    this.detachFns.push(this.props.sdk.entry.onSysChanged(this.onSysChanged))
+
+    this.props.sdk.window.startAutoResizer()
+  }
+
+  componentWillUnmount = () => {
+    this.detachFns.forEach(detach => detach())
+  }
+
+  constructState = () => {
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: remove

File: samples/gatsby-preview/src/index.jsx [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+      webhookUrl:
+        "https://backends.ctffns.net/gatsby-preview-proxy/" + projectId,
+      debounceInterval: null
+    }
+  }
+
+  componentDidMount = () => {
+    this.detachFns = []
+
+    this.detachFns.push(this.props.sdk.entry.onSysChanged(this.onSysChanged))
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: No need for a list of listeners, just do:

`this.off = this.props.sdk.entry.onSysChanged(this.onSysChanged)`

and

`this.off()`

in `componentWillUnmount`

File: samples/gatsby-preview/src/index.jsx [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+  refreshGatsbyPreview = () => {
+    clearInterval(this.state.debounceInterval)
+
+    fetch(this.state.webhookUrl, {
+      method: "POST",
+      headers: { "Content-Type": "application/json" },
+      body: JSON.stringify({})
+    })
+      .then(res => res.json())
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: You don't rely on the response so there's no need to read JSON out of it.

File: samples/gatsby-preview/src/index.jsx [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+      .then(
+        function(jsonResp) {
+          this.props.sdk.notifier.success("Gatsby Preview Updated!")
+        }.bind(this),
+        function(error) {
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: you can do `error => this.props.sdk.notifier.error(error)` so you don't have to bind.

File: samples/gatsby-preview/src/index.jsx [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+      return
+    }
+    if (this.state.debounceInterval) {
+      clearInterval(this.state.debounceInterval)
+    }
+    this.state.debounceInterval = setInterval(this.refreshGatsbyPreview, 1000)
+  }
+
+  refreshGatsbyPreview = () => {
+    clearInterval(this.state.debounceInterval)
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: How about the approach with `this.updated`? You setting and clearing intervals all the time is hard to follow.

File: samples/gatsby-preview/src/index.jsx [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    const { contentTypeSlug, isAutoUpdate } = parameters.instance
+
+    this.state = {
+      projectId,
+      projectUrl:
+        "https://" + projectId + "-preview.gtsb.io/" + contentTypeSlug + "/",
+      isAutoUpdate: isAutoUpdate,
+      webhookUrl:
+        "https://backends.ctffns.net/gatsby-preview-proxy/" + projectId,
+      debounceInterval: null,
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: Don't keep in state. See `this.detachFn`

File: samples/gatsby-preview/src/index.jsx [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+  onSysChanged = () => {
+    if (!this.state.isAutoUpdate) {
+      return
+    }
+    this.state.isUpdated = true
+  }
+
+  refreshGatsbyPreview = () => {
+    if (!this.state.isUpdated) {
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: You should also check `isAutoUpdate`

File: samples/gatsby-preview/src/index.jsx [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    }
+    
+    this.state.isUpdated = false
+
+    fetch(this.state.webhookUrl, {
+      method: "POST",
+      headers: { "Content-Type": "application/json" },
+      body: JSON.stringify({})
+    })
+      .then(success => this.props.sdk.notifier.success("Gatsby Preview Updated!"), error => this.props.sdk.notifier.error(error))
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: Can you break lines here?

```
then(
  x => ...,
  y => ...
)
```

File: samples/gatsby-preview/src/index.jsx [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+
+  render = () => {
+    return (
+      <div className="extension">
+        <div className="flexcontainer">
+          <Forma36.Button buttonType="positive" onClick={this.openPreviewTab}>
+            Open preview
+          </Forma36.Button>
+
+          {/* <Forma36.FieldGroup>
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: Do we tackle it?

File: samples/gatsby-preview/src/index.jsx (hunk 11 above)
Comment (floelhoeffel) [reply samples/gatsby-preview/src/index.jsx]: this method is also invoked with a button, that is why I am not checking

File: samples/gatsby-preview/src/index.jsx (hunk 13 above)
Comment (floelhoeffel) [reply samples/gatsby-preview/src/index.jsx]: removed

File: samples/gatsby-preview/src/index.jsx [hunk 14]
```diff
@@ -0,0 +1,144 @@
+class App extends React.Component {
+  constructor(props) {
+    super(props)
+
+    const parameters = this.props.sdk.parameters
+    const projectId = parameters.installation.projectId
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: You can actually do `{ parameters }` and `{ projectId }` here, as you do below.

File: samples/gatsby-preview/src/index.jsx [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+
+    const { contentTypeSlug, isAutoUpdate } = parameters.instance
+
+    this.state = {
+      projectId,
+      projectUrl:
+        "https://" + projectId + "-preview.gtsb.io/" + contentTypeSlug + "/",
+      isAutoUpdate: isAutoUpdate,
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: no need to repeat

File: samples/gatsby-preview/src/index.jsx [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+    if (this.debounceInterval) {
+      clearInterval(this.debounceInterval)
+    }
+
+    fetch(this.state.webhookUrl, {
+      method: "POST",
+      headers: { "Content-Type": "application/json" },
+      body: JSON.stringify({})
+    }).then(
+      success => this.props.sdk.notifier.success("Gatsby Preview Updated!"),
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: isn't it too frequent to show a notification like this? Shouldn't we just indicate in the extension somewhere?

File: samples/gatsby-preview/src/index.jsx [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+      error => this.props.sdk.notifier.error(error)
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: I'd just hardcode some string ("failed to update"). Error will be an object so you'll get an notification with `[object Object]`

File: samples/gatsby-preview/src/index.jsx [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+    )
+  }
+
+  openPreviewTab = () => {
+    window.open(
+      this.state.projectUrl + this.props.sdk.entry.fields.slug.getValue()
+    )
+  }
+  onChange = e => {
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: I'd name it `onAutoUpdateChange`

File: samples/gatsby-preview/src/index.jsx [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+    this.setState({ isAutoUpdate: e.target.value === "yes" }, () => {
+      console.log(this.state.isAutoUpdate)
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: drop the log and the commented out line below

File: samples/gatsby-preview/src/index.jsx [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+          <Forma36.Button
+            size="small"
+            buttonType="muted"
+            onClick={this.refreshGatsbyPreview}
+          >
+            Update Preview
+          </Forma36.Button>
+
+          <div
+            style={{
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: You're mixing inline styles and styles defined in `index.html`. I'd pick one option and stick to it.

File: samples/gatsby-preview/src/index.jsx [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+              display: "flex",
+              flexDirection: "row",
+              alignItems: "center",
+              justifyContent: "flex-end"
+            }}
+          >
+            <Forma36.Paragraph
+              extraClassNames=""
```
Comment (jelz) [inline samples/gatsby-preview/src/index.jsx]: skip

File: samples/gatsby-preview/build/index.html [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+    <link
+      rel="stylesheet"
+      href="//unpkg.com/@contentful/forma-36-fcss/dist/styles.css"
+    />
+    <style type="text/css">
+      body {
+        margin: 0;
+        padding: 0;
+      }
+      .full-width {
```
Comment (jelz) [inline samples/gatsby-preview/build/index.html]: I don't think it's used.

File: samples/gatsby-preview/src/index.jsx (hunk 16 above)
Comment (floelhoeffel) [reply samples/gatsby-preview/src/index.jsx]: yeah be too noisy, I still feel it is kind of cool so I'd propose to leave it as is for now

Comment (jelz) [reply samples/gatsby-preview/src/index.jsx]: ok, cool factor > everything else ;)

**Comment by floelhoeffel** [comment]
PTAL @jelz

**Comment by jelz** [comment]
@floelhoeffel looks good. Do we want to use the micro-backend for that or waiting for CORS?

**Comment by floelhoeffel** [comment]
Thing is, I dont have any commitments / timelines for CORS support. Should we swap the microbackend with webtask or something for now?

Or should I add a disclaimer in a comment somewhere instead?
